ocl/oclapi/filters.py

```diff
@@ -152,21 +153,24 @@ def _filter_queryset(self, request, queryset, view, sqs):
 
 class HaystackSearchFilter(BaseHaystackSearchFilter):
     def filter_queryset(self, request, queryset, view):
-        return self._filter_queryset(request, queryset, view, SearchQuerySet())
+        return self._filter_queryset(request, queryset, view, RelatedSearchQuerySet())
```

Comment:
I explicitly wanted to avoid RelatedSearchQuerySet due to performance issues we had with that in the past. I see you also introduced a call to load_all()... Do you have an understanding of what it entails? Is it loading all resources from solr query?

Reply:
@Karuhanga, ping. Just wanted to follow up on this.

Reply:
Yes, the Haystack docs have warning about this(https://django-haystack.readthedocs.io/en/v2.0.0/searchqueryset_api.html#relatedsearchqueryset) and I did consider it.
However, I failed to make the same logic function with a `SQ` query hence making that tradeoff.

Reply:
> Is it loading all resources from solr query, Does it respect pagination at all?
Everything before your first offset
> Does it build items fetching them from mongodb as well?
Not sure I understand this one

Reply:
> Does it build items fetching them from mongodb as well?
> 
> Not sure I understand this one

Typically a query is first executed against solr and then against mongodb to build a proper model. The question is, if load_all does that or it is just solr.

Reply:
In other words how many queries are executed against solr and mongob in sqs = sqs.load_all().load_all_queryset(view.model, queryset)
Is it 4 or 2 as usual?

Reply:
It should still be 2.
My mental model was previously it was query solr and use IDs to query the dB. Now it is query solr and use the IDs to restrict results from this query.
I think the only difference is we fetch all objects less than offset 1 from the dB.

Reply:
Now that you mention it though, I think I used the related sqs for everything and not only collections and sources.

Reply:
The thing is that you call both load_all() and load_all_queryset(). Not sure why you need to do it twice, thus my concern it executes more queries than needed.

We should definitely not call load_all/load_all_queryset, if not absolutely needed. Not sure, if we can discover, if queryset has any filters applied and only then call load_all_queryset... If not, then we definitely need to limit the use of it to collection/sources. It's still problematic since most of the time we are fine with using only solr stored values. Ideally, I'd see why you couldn't make it work with solr query in the first place.

Reply:
> We should definitely not call load_all/load_all_queryset, if not absolutely needed

I am going to do a check to confirm and revert here, but I believe `load_all` merely sets a flag and does not immediately run any queries.

> if queryset has any filters applied

In this case, that will always be yes since it's used for permissioning

> It's still problematic since most of the time we are fine with using only solr stored values. Ideally, I'd see why you couldn't make it work with solr query in the first place.

I agree. I am going to give this one more look later today, but if you reset to this commit(2d6b16026055b3e646fb40114f4e20adcb68f6ed) and run those tests, I added in that particular commit, I failed to find a way to achieve that functionality with solr values alone.

Reply:
One more thought,
If we do decide to move this to Haystack queries only, does this mean that we'll stop doing all `queryset` modifications in views?

Reply:
https://staging.openconceptlab.org/search/?q= got much slower after deploying recent changes. Not sure, if it is related, but it seems so.

We cannot rely entirely on solr queries, but we should rely as much as possible on them for full text searches. My understanding is that load_all_queryset is more performant with less results from solr. We should either auto-discover, if queryset has any filters or enable that feature explicitly, before calling load_all_queryset. We should also revisit privileges code to apply restrictions in one place (either via solr or mongo) as the last query step.

Reply:
Sorry I missed this @rkorytkowski, I'll go through my notifications more regularly going forward.
I'll pick up on this;
`revisit privileges code to apply restrictions in one place (either via solr or mongo) as the last query step`.
You'll actually notice I tried to separate the filter logic from the permissions logic.
The problem with using either of the two solely was that when you had multiple filter criteria, for example filter by orgs based on the url and then by whether someone is a member of the org for permissions, that failed, as in literally crashed.